crates/sui-core/src/histogram.rs

```diff
@@ -0,0 +1,389 @@
+        // data points being dropped way before it will produce any measurable memory pressure
+        #[allow(clippy::disallowed_methods)]
+        let (sender, receiver) = mpsc::unbounded_channel();
+        let reporter = HistogramReporter {
+            gauge,
+            sum,
+            count,
+            percentiles,
+            known_labels: Default::default(),
+        };
+        let reporter = Arc::new(Mutex::new(reporter));
+        let collector = HistogramCollector {
+            reporter,
+            channel: receiver,
+        };
+        Handle::current().spawn(collector.run());
+        Self { channel: sender }
+    }
+
+    pub fn with_label_values(&self, labels: &[&str]) -> Histogram {
+        let labels = labels.iter().map(ToString::to_string).collect();
+        let labels = HistogramLabelsInner::new(labels);
+        Histogram {
+            labels,
+            channel: self.channel.clone(),
+        }
+    }
+}
+
+impl HistogramLabelsInner {
+    pub fn new(labels: Vec<String>) -> HistogramLabels {
+        // Not a crypto hash
+        let mut hasher = DefaultHasher::new();
+        labels.hash(&mut hasher);
+        let hash = hasher.finish();
+        Arc::new(Self { labels, hash })
+    }
+}
+
+impl PartialEq for HistogramLabelsInner {
+    fn eq(&self, other: &Self) -> bool {
+        self.hash == other.hash
+    }
+}
+
+impl Eq for HistogramLabelsInner {}
+
+impl Hash for HistogramLabelsInner {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.hash.hash(state)
+    }
+}
+
+impl Histogram {
+    pub fn report(&self, v: Point) {
+        if self.channel.send((self.labels.clone(), v)).is_err() {
+            unreachable!("Histogram reporting task do not stop when Histogram instance exists");
+        }
+    }
+
+    pub fn start_timer(&self) -> HistogramTimerGuard {
+        HistogramTimerGuard {
+            histogram: self,
+            start: Instant::now(),
+        }
+    }
+}
+
+impl HistogramCollector {
+    pub async fn run(mut self) {
+        let mut deadline = Instant::now();
+        loop {
+            // We calculate deadline here instead of just using sleep inside cycle to avoid accumulating error
+            deadline += Duration::from_secs(1);
+            if self.cycle(deadline).await.is_err() {
+                return;
+            }
+        }
+    }
+
+    async fn cycle(&mut self, deadline: Instant) -> Result<(), ()> {
+        let mut labeled_data: HashMap<HistogramLabels, Vec<Point>> = HashMap::new();
+        let mut timeout = tokio::time::sleep_until(deadline).boxed();
+        loop {
+            tokio::select! {
+             _ = &mut timeout => {
+                break;
+            },
+                point = self.channel.recv() => {
+                    if let Some((label, point)) = point {
+                        let values = labeled_data.entry(label).or_default();
+                        values.push(point);
```

Comment:
Should we also break out of the loop based on the count (along with time duration) and flush buffer sooner if we've accumulated enough samples?

Reply:
Yes, I think it's a good idea to limit number of points accumulated. We can't flush data earlier as prometheus collects data on fixed schedule, but we should drop extra points